libs/StateSync.Managed/src/NetworkManager.cs

```diff
@@ -0,0 +1,222 @@
+            }
+        }
+
+        public struct ListenerWrapper : NetworkListener
+        {
+            public unsafe ListenerWrapper(void* listener) { this.listener = listener; }
+
+            unsafe void* listener;
+
+            public unsafe void OnMessage(InternedBlobRef senderConnectionString, ReadOnlySpan<byte> message)
+            {
+                fixed (byte* bytes = message)
+                {
+                    NetworkListener_OnMessage(listener, senderConnectionString.handle, bytes, message.Length);
+                }
+            }
+        }
+
+        public NetworkManagerPInvokeWrapper(NetworkManager manager)
+        {
+            manager_ = manager;
+        }
+
+        public unsafe delegate void GetConnectionDelegate(
+            IntPtr connectionString,
+            IntPtr managerGCHandle,
+            void* networkManagerWrapper,
+            void* resultLocation);
+
+
+        public unsafe delegate void PollMessageDelegate(
+            IntPtr managerGCHandle,
+            void* listener);
+
+        public unsafe delegate void SendMessageDelegate(
+            IntPtr connectionGCHandle,
+            byte* messageBegin,
+            int messageSize);
+
+        /// <summary>
+        /// Invokes <see cref="NetworkManager.GetConnection(InternedBlobRef)"/>
+        /// and writes std::shared_ptr to the C++ wrapper object to the result to the location
+        /// pointed by <paramref name="resultLocation"/>. The wrapper is either obtained from
+        /// the cache of wrappers, or created on demand.
+        /// </summary>
+        public static unsafe void InvokeGetConnection(
+            IntPtr connectionString,
+            IntPtr managerGCHandle,
+            void* networkManagerWrapper,
+            void* resultLocation)
+        {
+            var managerWrapper =
+                GCHandle.FromIntPtr(managerGCHandle).Target as NetworkManagerPInvokeWrapper;
+            try
+            {
+                var connection = managerWrapper.manager_.GetConnection(
+                    new InternedBlobRef(connectionString));
+                lock (managerWrapper.connectionWrappersCache_)
+                {
+                    NetworkConnectionWeakPtr weakWrapper;
+                    if (!managerWrapper.connectionWrappersCache_.TryGetValue(connection, out weakWrapper))
+                    {
+                        weakWrapper = new NetworkConnectionWeakPtr();
+                        managerWrapper.connectionWrappersCache_.Add(connection, weakWrapper);
+                    }
+                    weakWrapper.InitConnectionWrapper(connection, networkManagerWrapper, connectionString, resultLocation);
+                }
+            }
+            catch
+            {
+                // the result pointed by resultLocation on the caller's side will stay empty.
+            }
+        }
+
+        /// <summary>
+        /// Invokes <see cref="NetworkConnection.SendMessage(ReadOnlySpan{byte})"/>
```

Comment:
nit: the comment is copy-and-paste from the method below, but the method is different 😄